src/Disks/S3/DiskS3.cpp

```diff
@@ -996,15 +995,18 @@ void DiskS3::restoreFileOperations(const RestoreInformation & restore_informatio
 
             LOG_TRACE(log, "Move directory to 'detached' {} -> {}", path, detached_path);
 
-            fs::path from_path = fs::path(metadata_path) / path;
-            fs::path to_path = fs::path(metadata_path) / detached_path;
+            fs::path from_path = fs::path(path);
+            fs::path to_path = fs::path(detached_path);
             if (path.ends_with('/'))
                 to_path /= from_path.parent_path().filename();
             else
                 to_path /= from_path.filename();
-            fs::create_directories(to_path);
-            fs::copy(from_path, to_path, fs::copy_options::recursive | fs::copy_options::overwrite_existing);
-            fs::remove_all(from_path);
+
+            /// to_path may exist and non-empty in case for example abrupt restart, so remove it before rename
+            if (metadata_disk->exists(to_path))
+                metadata_disk->removeRecursive(to_path);
+
+            metadata_disk->moveDirectory(from_path, to_path);
```

Comment:
`to_path` is `detached_path`, and `detached_path` always ends with `detached/`.
Now we remove `to_path` if it exists, then to path must be `non-existing directory, not ending with a directory separator, and whose parent directory exists`, but it actually ends with a directory separator and we will get an error.. So we need to remove it.

Reply:
yes, ok.